# VirMAP

The full VirMAP source code will be released upon publication. The manuscript describing the pipeline is currently under review. 

Working title, authors, and abstract: 

VirMAP, towards maximal viral information recovery from non-targeted sequencing
 
Nadim J Ajami1,2, Ω, Matthew C. Wong1,2, Ω, Matthew C. Ross1,2, Richard E. Lloyd2, Joseph F. Petrosino1,2.
 
1 Alkek Center for Metagenomics and Microbiome Research, and 2­ Department of Molecular Virology and Microbiology, Baylor College of Medicine, Houston, Texas.
Ω ­Authors contributed equally to this work.
 
Corresponding author:
 
Nadim J. Ajami
nadim.ajami@bcm.edu
1 Baylor Plaza, Rm 808ED, Houston, Texas, USA 77030
 
Abstract
 
Accurate classification of the human virome is critical to fully understand the role viruses play in health and disease. This implies the need for sensitive, specific, and practical pipelines that return precise outputs while still enabling case-specific post hoc analysis. Viral taxonomic characterization from metagenomic data presents many challenges including codon degeneracy, rapid mutation rate, and shared protein homology across distant taxa. VirMAP accurately assigns strain-level viral taxonomies by coherently sorting, mapping, and merging nucleotide and protein information. VirMAP achieves this using novel clustering and classification algorithms that merges nucleotide and protein information to taxonomic classify viral reconstructions independent of genome coverage or read overlap. VirMAP was created to address commonly cited caveats regarding viral classification and was validated using viral mock communities. This novel approach offers new opportunities to enhance metagenomic studies seeking to define virome-host interactions, improve biosurveillance capabilities, and strengthen molecular epidemiology reporting.
